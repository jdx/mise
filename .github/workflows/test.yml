name: test

on:
  push:
    branches: ["*"]
    tags: ["v*"]
  pull_request:
    branches: ["main"]

jobs:
  test:
    runs-on: ubuntu-22.04
    timeout-minutes: 10
    steps:
      - name: Checkout
        uses: actions/checkout@v3
      - uses: jdxcode/rtx-action@v1
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      - run: mkdir -p ~/.local/share/rtx/plugins
      - run: ln -s $PWD ~/.local/share/rtx/plugins/python
      - run: rtx i python@latest
      - run: rtx x python@latest -- python -V
