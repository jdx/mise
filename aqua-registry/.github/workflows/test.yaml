---
name: test
on:
  pull_request:
    branches: [main]
concurrency:
  group: ${{ github.workflow }}--${{ github.ref }}
  cancel-in-progress: true
permissions: {}
jobs:
  test:
    uses: ./.github/workflows/workflow_call_test.yaml
    permissions:
      contents: read
      pull-requests: read
  status-check:
    runs-on: ubuntu-24.04
    # https://github.com/aquaproj/aqua-registry/pull/38449#issuecomment-3038859524
    if: always() && (contains(needs.*.result, 'failure') || contains(needs.*.result, 'cancelled'))
    timeout-minutes: 10
    permissions: {}
    needs:
      - test
    steps:
      - run: exit 1
