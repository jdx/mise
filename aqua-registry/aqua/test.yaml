---
# aqua - Declarative CLI Version Manager
# https://aquaproj.github.io/
checksum:
  enabled: true
registries:
  - name: standard
    type: local
    path: ../registry.yaml
packages:
