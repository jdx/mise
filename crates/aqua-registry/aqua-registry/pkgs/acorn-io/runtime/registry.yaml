# yaml-language-server: $schema=https://raw.githubusercontent.com/aquaproj/aqua/main/json-schema/registry.json
packages:
  - type: github_release
    repo_owner: acorn-io
    repo_name: runtime
    aliases:
      - name: acorn-io/acorn
    description: A simple application deployment framework for Kubernetes
    asset: acorn-{{.Version}}-{{.OS}}-{{.Arch}}.{{.Format}}
    files:
      - name: acorn
    format: tar.gz
    replacements:
      darwin: macOS
    overrides:
      - goos: darwin
        asset: acorn-{{.Version}}-{{.OS}}-universal.{{.Format}}
      - goos: windows
        format: zip
    supported_envs:
      - darwin
      - linux
      - amd64
    checksum:
      type: github_release
      asset: checksums.txt
      algorithm: sha256
