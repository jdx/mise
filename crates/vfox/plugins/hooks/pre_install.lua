--- Returns some pre-installed information, such as version number, download address, local files, etc.
--- If checksum is provided, vfox will automatically check it for you.
--- @param ctx table
--- @field ctx.version string User-input version
--- @return table Version information
function PLUGIN:PreInstall(ctx)
    local version = ctx.version

    return {
        version = version,
    }
end
