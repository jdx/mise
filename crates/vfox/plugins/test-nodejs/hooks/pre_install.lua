--- Return the URL to download the tool
--- @param ctx table See /vfox/ctx.md#ctx-hooks for more information on ctx
--- @return table Version information
function PLUGIN:PreInstall(ctx)
    local version = ctx.version

    -- Return simple test data without runtime checks
    return {
        version = version,
        url = "file:///fake/nodejs/node-v" .. version .. ".tar.gz",
        sha256 = "fakehash" .. version,
    }
end