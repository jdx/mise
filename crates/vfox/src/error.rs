use mlua::Error as MLuaError;
use thiserror::Error;
use xx::XXError;

#[derive(Error, Debug)]
#[non_exhaustive]
pub enum VfoxError {
    #[error("{0}")]
    Error(String),
    #[error(transparent)]
    LuaError(#[from] MLuaError),
    #[error("serde_json")]
    SerdeJsonError(#[from] serde_json::Error),
    #[error(transparent)]
    XXError(#[from] XXError),
    #[error(transparent)]
    ReqwestError(#[from] reqwest::Error),
    #[error(transparent)]
    IoError(#[from] std::io::Error),
    #[error(transparent)]
    UrlParseError(#[from] url::ParseError),
}

pub type Result<T> = std::result::Result<T, VfoxError>;

impl From<String> for VfoxError {
    fn from(s: String) -> Self {
        VfoxError::Error(s)
    }
}

impl From<&str> for VfoxError {
    fn from(s: &str) -> Self {
        VfoxError::Error(s.to_string())
    }
}

#[macro_export]
macro_rules! error {
    ($($arg:tt)*) => {
        return Err(VfoxError::Error(format!($($arg)*)));
    };
}
