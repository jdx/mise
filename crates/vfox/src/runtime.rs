use crate::config::{arch, os};
use mlua::{UserData, UserDataFields};
use once_cell::sync::Lazy;
use std::path::PathBuf;
use std::sync::Mutex;

#[derive(Debug, Clone)]
pub(crate) struct Runtime {
    pub(crate) os: String,
    pub(crate) arch: String,
    pub(crate) version: String,
    pub(crate) plugin_dir_path: PathBuf,
}

static RUNTIME: Lazy<Mutex<Runtime>> = Lazy::new(|| {
    Mutex::new(Runtime {
        os: os(),
        arch: arch(),
        version: "0.6.0".to_string(), // https://github.com/version-fox/vfox/releases
        plugin_dir_path: PathBuf::new(),
    })
});

impl Runtime {
    pub(crate) fn get(plugin_dir_path: PathBuf) -> Runtime {
        let mut runtime = RUNTIME.lock().unwrap().clone();
        runtime.plugin_dir_path = plugin_dir_path;
        runtime
    }

    #[cfg(test)]
    pub(crate) fn set_os(os: String) {
        let mut runtime = RUNTIME.lock().unwrap();
        runtime.os = os;
    }

    #[cfg(test)]
    pub(crate) fn set_arch(arch: String) {
        let mut runtime = RUNTIME.lock().unwrap();
        runtime.arch = arch;
    }

    #[cfg(test)]
    pub(crate) fn reset() {
        let mut runtime = RUNTIME.lock().unwrap();
        runtime.os = os();
        runtime.arch = arch();
    }
}

impl UserData for Runtime {
    fn add_fields<F: UserDataFields<Self>>(fields: &mut F) {
        fields.add_field_method_get("osType", |_, t| Ok(t.os.clone()));
        fields.add_field_method_get("archType", |_, t| Ok(t.arch.clone()));
        fields.add_field_method_get("version", |_, t| Ok(t.version.clone()));
        fields.add_field_method_get("pluginDirPath", |_, t| Ok(t.plugin_dir_path.clone()));
    }
}
