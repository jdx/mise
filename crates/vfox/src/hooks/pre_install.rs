use mlua::prelude::LuaError;
use mlua::{FromLua, Lua, Value};

use crate::error::Result;
use crate::Plugin;

impl Plugin {
    pub async fn pre_install(&self, version: &str) -> Result<PreInstall> {
        debug!("[vfox:{}] pre_install", &self.name);
        let ctx = self.context(Some(version.to_string()))?;
        let pre_install = self
            .eval_async(chunk! {
                require "hooks/pre_install"
                return PLUGIN:PreInstall($ctx)
            })
            .await?;

        Ok(pre_install)
    }
}

#[derive(Debug)]
pub struct PreInstall {
    pub version: String,
    pub url: Option<String>,
    pub note: Option<String>,
    pub sha256: Option<String>,
    pub md5: Option<String>,
    pub sha1: Option<String>,
    pub sha512: Option<String>,
    // pub addition: Option<Table>,
}

impl FromLua for PreInstall {
    fn from_lua(value: Value, _: &Lua) -> std::result::Result<Self, LuaError> {
        match value {
            Value::Table(table) => {
                if !table.contains_key("version")? {
                    return Err(LuaError::FromLuaConversionError {
                        from: "table",
                        to: "PreInstall".into(),
                        message: Some("no version returned from vfox plugin".to_string()),
                    });
                }
                Ok(PreInstall {
                    version: table.get::<String>("version")?,
                    url: table.get::<Option<String>>("url")?,
                    note: table.get::<Option<String>>("note")?,
                    sha256: table.get::<Option<String>>("sha256")?,
                    md5: table.get::<Option<String>>("md5")?,
                    sha1: table.get::<Option<String>>("sha1")?,
                    sha512: table.get::<Option<String>>("sha512")?,
                    // addition,
                })
            }
            _ => panic!("Expected table"),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::hooks::pre_install::PreInstall;
    use crate::runtime::Runtime;
    use crate::Plugin;
    use std::string::ToString;
    use tokio::test;

    #[test]
    async fn dummy() {
        let pre_install = run("dummy", "1.0.1").await;
        assert_debug_snapshot!(pre_install);
    }

    #[test]
    async fn nodejs() {
        Runtime::set_os("linux".to_string());
        Runtime::set_arch("x64".to_string());
        let pre_install = run("nodejs", "20.0.0").await;
        assert_debug_snapshot!(pre_install);

        Runtime::set_os("macos".to_string());
        Runtime::set_arch("arm64".to_string());
        let pre_install = run("nodejs", "20.1.0").await;
        assert_debug_snapshot!(pre_install);

        Runtime::set_os("windows".to_string());
        Runtime::set_arch("x64".to_string());
        let pre_install = run("nodejs", "20.3.0").await;
        assert_debug_snapshot!(pre_install);

        Runtime::reset();
    }

    async fn run(plugin: &str, v: &str) -> PreInstall {
        let p = Plugin::test(plugin);
        p.pre_install(v).await.unwrap()
    }
}
