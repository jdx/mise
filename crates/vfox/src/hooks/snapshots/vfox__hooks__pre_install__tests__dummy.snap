---
source: src/hooks/pre_install.rs
expression: pre_install
---
PreInstall {
    version: "1.0.1",
    url: None,
    note: None,
    sha256: None,
    md5: None,
    sha1: None,
    sha512: None,
}
